### PR TITLE
Delivery tracking + badges

### PR DESCRIPTION
**Pull request by ftcvlad**
### Description of Changes

Ported from Swift

1. push delivery tracking
2. incrementing badges

### Breaking Changes

-   None

### Release Checklist

Prepare:

-   [x] Detail any breaking changes. Breaking changes require a new major version number
-   [x] Check `pod lib lint` passes

Bump versions in:

-   [x] `Sources/Kumulos+Stats.m`
-   [x] `KumulosSdkObjectiveC.podspec`
-   [x] `KumulosSdkObjectiveCExtension.podspec`
-   [x] All relevant build targets
-   [x] `README.md`

Release:

-   [ ] Squash and merge to master
-   [ ] Delete branch once merged
-   [ ] Create tag from master matching chosen version
-   [ ] Run `pod trunk push` to publish to CocoaPods

### CODE REVIEW

File: Sources/InApp/KSInAppHelper.m [hunk 1]
```diff
@@ -117,7 +116,7 @@ - (void) setupSyncTask {
 #pragma mark - State helpers
 
 -(NSString*) keyForUserConsent {
```
Comment (cgwyllie) [inline Sources/InApp/KSInAppHelper.m:118]: Could just factor this method out? (don't think is ported to swift)

File: Sources/Kumulos+Push.m [hunk 2]
```diff
@@ -222,6 +222,16 @@ - (NSString*) pushTokenFromData:(NSData*) deviceToken {
     return [token copy];
 }
 
+- (void) trackPushDelivery:(NSDictionary*) userInfo{
+    KSPushNotification* notification = [KSPushNotification fromUserInfo:userInfo];
+    if (nil == notification) {
```
Comment (cgwyllie) [inline Sources/Kumulos+Push.m:227]: Does this also need to guard id != 0?

File: Sources/Kumulos+Push.m [hunk 3]
```diff
@@ -265,21 +275,42 @@ void kumulos_applicationDidReceiveRemoteNotificationFetchCompletionHandler(id se
             [Kumulos.shared pushHandleOpenWithUserInfo:userInfo];
         }
     }
-
+    
     if ([userInfo[@"aps"][@"content-available"] intValue] == 1) {
+        NSNumber* newBadgeValue = [KumulosHelper getBadgeFromUserInfo:userInfo];
```
Comment (cgwyllie) [inline Sources/Kumulos+Push.m:280]: Control flow in here w.r.t. iOS 9 / 10 open/badge/content-available handling seems different to Swift?

https://github.com/Kumulos/KumulosSdkSwift/pull/58/files#diff-a40628d66e7a872b708cef4259773e6bR362

File: Sources/KumulosSDKExtension/KumulosNotificationService.m [hunk 4]
```diff
@@ -6,27 +6,39 @@
...
+    NSDictionary* custom = userInfo[@"custom"];
+    NSDictionary* data = custom[@"a"];
+    NSDictionary* msg = data[@"k.message"];
+    NSDictionary* msgData = msg[@"data"];
+    NSNumber* messageId = msgData[@"id"];
+    
+    [self maybeSetBadge:bestAttemptContent userInfo:userInfo];
+    [self trackDeliveredEvent:userInfo notificationId: messageId];
     
-    if  (userInfo && userInfo[@"custom"] && userInfo[@"custom"][@"a"] && userInfo[@"custom"][@"a"][@"k.message"] && userInfo[@"custom"][@"a"][@"k.buttons"]) {
```
Comment (cgwyllie) [inline Sources/KumulosSDKExtension/KumulosNotificationService.m:20]: Removing these guards as above may cause nil ptrs if user passes a local notification not conforming to our shape in here?

Maybe better to use the model factory method and test id != nil before proceeding?

File: Sources/SessionHelper.m [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+//
+//  Created by Vladislav Voicehovics on 27/03/2020.
+//
+
+#import <Foundation/Foundation.h>
+#import "SessionHelper.h"
+#import "Kumulos+Protected.h"
+#import "KumulosEvents.h"
+
+@interface SessionHelper ()
```
Comment (cgwyllie) [inline Sources/SessionHelper.m:13]: We should prefix the class names `KS..`

File: Sources/Shared/AnalyticsHelper.m [hunk 6]
```diff
@@ -84,82 +85,113 @@ @implementation AnalyticsHelper
...
 #pragma mark - Initialization
 
-- (instancetype _Nullable) initWithKumulos:(Kumulos *)kumulos {
+- (instancetype _Nullable) initWithApiKey:(NSString*)apiKey withSecretKey:(NSString*)secretKey {
     if (self = [super init]) {
-        self.kumulos = kumulos;
-        self.startNewSession = YES;
-        self.sessionIdleTimer = nil;
-        self.bgTask = UIBackgroundTaskInvalid;
+#if TARGET_OS_IOS
```
Comment (cgwyllie) [inline Sources/Shared/AnalyticsHelper.m:90]: Guard necessary? Does this only include in targets for iOS?

File: Sources/Shared/KSAppGroupsHelper.m [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        NSBundle* mainBundle = [[NSBundle alloc] initWithURL:url];
+        if (mainBundle != nil){
+            targetBundle = mainBundle;
+        }
+        else{
+            NSLog(@"AppGroupsHelper: Error, could not obtain main bundle from extension!");
+        }
+    }
+    
+    //FIXME: test correct group in app/in extension
```
Comment (cgwyllie) [inline Sources/Shared/KSAppGroupsHelper.m:39]: Relevant/fixed?

File: Sources/Shared/KumulosHelper.m [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+#import "KumulosUserDefaultsKeys.h"
+#import "KSKeyValPersistenceHelper.h"
+
+
+@implementation KumulosHelper
+
+static NSString* _Nonnull const userIdLocker = @"";
+
++ (NSString*) installId {
+    @synchronized (self) {
```
Comment (cgwyllie) [inline Sources/Shared/KumulosHelper.m:19]: Is synchronising over static self still correct, or should it be similar to using the userIdLocker?

File: Sources/Shared/KumulosUserDefaultsKeys.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
+//
+//  KumulosUserDefaultsKeys.h
+//  KumulosSDK iOS
+//
+//  Created by Vladislav Voicehovics on 27/03/2020.
+//
+
+#import <Foundation/Foundation.h>
+
+extern NSString* _Nonnull const KumulosApiKey;
```
Comment (cgwyllie) [inline Sources/Shared/KumulosUserDefaultsKeys.h:10]: Naming q, worth effort of renaming w/ common prefix like `KSPrefsKey...` or `KumulosKey...` etc.?

File: Sources/Kumulos+Push.m (hunk 3 above)
Comment (ftcvlad) [reply Sources/Kumulos+Push.m:280]: track delivery, set badge if:
a) content-available=0 && ios 9
b) content-available=1

so, both equivalent i think. I haven't changed control flow, it was different in objc. Shall make as in swift?

File: Sources/Shared/KSAppGroupsHelper.m (hunk 7 above)
Comment (ftcvlad) [reply Sources/Shared/KSAppGroupsHelper.m:39]: fixed

File: Sources/Shared/AnalyticsHelper.m (hunk 6 above)
Comment (ftcvlad) [reply Sources/Shared/AnalyticsHelper.m:90]: true

File: Sources/SessionHelper.m (hunk 5 above)
Comment (ftcvlad) [reply Sources/SessionHelper.m:13]: fixed

File: Sources/Kumulos+Push.m (hunk 2 above)
Comment (ftcvlad) [reply Sources/Kumulos+Push.m:227]: id can be 0 in swift, dont see it 0 in objc

File: Sources/Shared/KumulosHelper.m (hunk 8 above)
Comment (ftcvlad) [reply Sources/Shared/KumulosHelper.m:19]: i thought you just use object/self as a lock , doesn't matter. this method copied from Kumulos.m

File: Sources/Shared/KumulosUserDefaultsKeys.h (hunk 9 above)
Comment (ftcvlad) [reply Sources/Shared/KumulosUserDefaultsKeys.h:10]: like KSPrefsKey

File: Sources/KumulosSDKExtension/KumulosNotificationService.m (hunk 4 above)
Comment (ftcvlad) [reply Sources/KumulosSDKExtension/KumulosNotificationService.m:20]: this was actually copied from swift. Should do the same there as well i suppose?

File: Sources/Kumulos+Push.m (hunk 3 above)
Comment (cgwyllie) [reply Sources/Kumulos+Push.m:280]: No need, I'll try reading from the file, not the diff, was just finding hard to follow.

File: Sources/KumulosSDKExtension/KumulosNotificationService.m (hunk 4 above)
Comment (cgwyllie) [reply Sources/KumulosSDKExtension/KumulosNotificationService.m:20]: Ok, yes some kind of guard might be appropriate then.

File: Sources/Shared/KumulosHelper.m (hunk 8 above)
Comment (cgwyllie) [reply Sources/Shared/KumulosHelper.m:19]: OK yep, should be fine, it will use the KumulosHelper.class for locking over. So long as not in a subclass, this should be fine: https://vombat.tumblr.com/post/83074022251/synchronizing-around-a-class

File: Sources/Kumulos+Push.m (hunk 3 above)
Comment (cgwyllie) [reply Sources/Kumulos+Push.m:280]: Concur, flow equivalent & as described, my bad reading of the diff.
